Woo: Add v4 revenue stats api support

anitaa1990 opened the pull request:
Fixes #1301. This PR adds support for the v4 stats api.  

### Changes:
- [x] Add new model `WCOrderStatsV4Model` for storing v4 api data
- [x] Add migration logic to create a new table for `WCOrderStatsV4Model` in WellSql
- [x] Add util class `WCStatsV4SqlUtils` with methods to insert/update and fetch `WCOrderStatsV4Model` from local db.
- [x] Added v4 api path and revenue stats endpoint.
- [x] Add method in the `OrderStatsRestClient` for wc-admin v4 endpoint and a new api response object `RevenueStatsApiResponse` to parse the v4 api response.
- [x] Add separate `WCStatsAction` for fetching revenue stats and handling fetched revenue stats
- [x] Add separate `OnWCStatsV4Changed` for listening to changes in the v4 stats api.
- [x] Add methods in `WcStatsStore` to fetch stats from v4 endpoint and store it in cache
- [x] Add separate `OrderApiUnit` value to support `HOURS`
- [x] Add mocked tests, sql unit tests and utils tests.
- [x] Add a new button in in the example app to fetch the revenue/order stats from v4 api

### Some Gotachas
- The revenue api has a parameter called `per_page` which can be used to get the maximum number of items in the response object. [The default value of this in the api is 10](https://github.com/woocommerce/woocommerce-admin/blob/6fade24afd9335469c85dc8618421276bd407a43/includes/api/class-wc-admin-rest-reports-revenue-stats-controller.php#L329). For the current month tab, we would need to fetch data for each of the 30 or 31 days in a month. So this value needs to be included when requesting for data from the api. But for some reason, if we send a static `per_page` value of any number, the api is not providing refreshed data. To reproduce this issue:
    - Pull the changes from this branch.
    - Change the value of `per_page` to a static value [in this line](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/ea38709d396b2bcf74e3644f96bb6dd59b763f98/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/orderstats/OrderStatsRestClient.kt#L160),
    - Build the app and click on the `Fetch current day Revenue Stats (Forced)` button in the example app.
    - Notice the stats value displayed.
    - Create a new order from a test site. Mark the order as completed.
    - Click on the `Fetch current day Revenue Stats (Forced)` button in the example app again.
    - Notice the revenue value has not changed. 

I tested the api directly using Postman and setting a random value seems to fix the issue. So, I've added logic, at the moment, to pass a random value from ~~1~~ 31 to 100 when requesting for data from the api. This random number will only be generated if `forced=true`. This random value generated will be stored locally in `SharedPreferences` and retrieved as the `per_page` param when `forced=false`. This way the request will be identical and caching will work on the FluxC side.

- Unlike the wp.com API, the v4 stats api does not provide any information regarding the currency. AFAIK, the `wc-admin` is using the currency specified in the Stats settings. So we would need to change the logic in the woo app to get currency value from the Site Settings. cc @ctarda since you are working on the iOS implementation.

- ~~The v4 API, provides an option to pass an ISO formatted start date and end date. So, rather than handling the logic of passing the date object based on the `StatsGranularity`, the current implementation requires that a start date be passed when requesting stats info from FluxC. This would mean the Woo app would need to pass the start date. The end date, by default is the current date. I thought this would be a better option, if we need to include support for custom date ranges, in the future. When the start date is passed to FluxC, the date is formatted according to the site's timezone before requesting for information.~~ 
I've modified the above logic and made the `startDate` as optional in order to align with the current implementation. This is because according to the new designs, when:
    - Today tab is selected: OrderStatsApiUnit field passed to the API should be HOUR. Since we would need to display the stats for each hour of the day.
    - This week tab is selected: OrderStatsApiUnit field passed to the API should be DAY. Since we would need to display the stats for each day of the week.
     - This month tab is selected: [OrderStatsApiUnit] field passed to the API should be [DAY]. Since we would need to display the stats for each day of the month.
     - This year tab is selected: OrderStatsApiUnit field passed to the API should be MONTH. Since we would need to display the stats for each month of the year.
 So the `OrderStatsApiUnit` would be the same for This week and This month. So I’ve modified the logic to store the `StatsGranularity` to the local cache instead of the `OrderStatsApiUnit` so that we would be able to differentiate between the data for each tab in the UI. 

### Testing:
- [x] Run tests in `DateUtilsTest`, `WCStatsStoreTest`, `WCStatsV4SqlUtilsTest`, `MockedStack_WCStatsTest`.
- [x] Verify that app upgrade is working correctly.
- [x] Verify that a new install is working correctly.
- [x] Testing the v4 api changes from the example app using a site without `wc-admin`. An error message should be displayed stating `Plugin not available`.
- [x] Testing the v4 api changes from the example app using a site with `wc-admin`.
- [x] Testing the current stats functionality is working as expected.

### Screenshots:
I have added the v4 changes in a separate fragment and also included a Site picker so we can test the stats changes for WooCommerce sites with and without `wc-admin` and non WooCommerce sites.
<img width="300" src="https://user-images.githubusercontent.com/22608780/61345358-5eb26b00-a872-11e9-9786-0442dd67f7a8.gif">

cc @AmandaRiu @nbradbury

anitaa1990 commented:
@AmandaRiu @nbradbury PR is ready for review. FYI, I am still testing for potential bugs due to timezone issues but would love to get your review on this, in the meantime. Thank you!!!!

AmandaRiu commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/orderstats/OrderStatsRestClient.kt:
```diff
@@ -52,6 +59,13 @@ class OrderStatsRestClient(
+    // so if we need to get data for an entire month without pagination, the per_page value should be 30 or 31.
+    // But, for some reason, if we keep the per_page value static, the api is not providing refreshed data
+    // when a new order is completed. So for now, keeping the per_page value as random between 1 to 100.
+    private val STATS_DEFAULT_PER_PAGE = 100
```
This is a big problem for a couple reasons:
- it nullifies the caching on our end because caching will only work if the request is identical to what is being cached. Changing the `per_page` changes the request, so a new request is being submitted every time regardless of the value of `forced`.
- the current workaround generates a random number between 0 and 100, which means that if the generated number is below the total breakdown for the selected interval (for example: interval = hours, `per_page` = 6), then we're missing all the interval data needed to populate our charts in the app. 

I kept retrying forced with the `per_page` set to 50 and it took about 5 minutes for the API to return fresh data which tells me they are caching requests on the API side as well. So there are a couple questions to dig into with the API folks:
- Does the v4 API provide a parameter to request fresh data? 
- How long do they cache the requests?
We may need to open a ticket on the API repo to request they add a way to request fresh data.

For now we can work around this issue by implementing the following logic:
- Since the interval breakdown is so important, and the range is limited for our scope, I recommend choosing a random number between 25 and 100. 
- Only generate a new random number if `forced=true`. For this to work we'll need to store the last generated random number to reuse as the `per_page` when `forced=false`. This way the request will be identical and caching will work on our end.

anitaa1990 replied:
Great catch!! Thank you!! I made modifications to the code based on your suggestions in d8401b7. I also added a fix to generate a random value if the sharedPreferences value is empty in 0979fa1 . I’ll get in touch with the API team on what is the logic in the backend.

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt:
```diff
@@ -251,4 +253,23 @@ object DateUtils {
+     * returns a formatted date that accounts for the site's timezone setting,
+     * in the format yyy-MM-ddThh:mm:ss with the time always set to the start of the [dateString]
+     */
+    fun getStartDateForSite(site: SiteModel, dateString: String): String {
```
Done in 63c665e

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt:
```diff
@@ -251,4 +253,23 @@ object DateUtils {
+     * in the format yyy-MM-ddThh:mm:ss with the time always set to the end of the
+     * current date
+     */
+    fun getEndDateForSite(site: SiteModel): String {
```
Done in 63c665e

anitaa1990 commented:
Thanks for the great review @AmandaRiu !
> I noticed that we use the current date as the end date for all the requests and at first glance that seems ok because we could always code the front end to fill in the future graph bars with 0's, but the I remembered hearing something about plugins that allow for scheduling orders or making recurring orders. This would mean we could possibly have orders in the future, but I'm not 100% certain of it. Definitely something we should look into. If that's the case we'll want to send an end date that covers the full range the interval represents.

Oh, I was not aware of this scenario. I replicated the functionality based on what I noticed in the `wc-admin`. The current date is passed as the end date param for all the tabs in `wc-admin`. Even for selecting custom date ranges, all the dates after the current date are disabled in the date selector. 
<img width="400" src="https://user-images.githubusercontent.com/22608780/61059081-0e707e80-a416-11e9-9f7b-732533160317.png">

anitaa1990 commented:
I made a couple of changes to the code over the weekend to reduce confusion regarding the naming of the new methods/classes and to align the functionality to be more aligned with the current implementation.
- I was not sure how to name things. I started with adding `WCRevenueStatsModel` but it was inherently confusing. So I modified all new items with the suffix v4. So the new model table will be `WcOrderStatsV4Model` and the new actions will be `FetchOrderStatsV4Payload` and `FetchOrderStatsV4ResponsePayload`. These changes were done in e4c6922
- The second change was to store the `StatsGranularity` to the local cache instead of `OrderStatsApiUnit`. This is because according to the new designs, when:
       *  `Today` tab is selected: `OrderStatsApiUnit` field passed to the API should be `HOUR`. Since we would need to display the stats for each hour of the day.
       *  `This week` tab is selected: `OrderStatsApiUnit` field passed to the API should be `DAY`. Since we would need to display the stats for each day of the week.
       *  `This month` tab is selected: [OrderStatsApiUnit] field passed to the API should be [DAY]. Since we would need to display the stats for each day of the month.
       *  `This year` tab is selected: `OrderStatsApiUnit` field passed to the API should be `MONTH`. Since we would need to display the stats for each month of the year.
So the `OrderStatsApiUnit` would be the same for `This week` and `This month`. So I’ve modified the logic to store the `StatsGranularity` to the local cache instead of the `OrderStatsApiUnit` so that we would be able to differentiate between the data for each tab in the UI. Does this make sense? The changes are added in 571764e 
- I modified the `FetchOrderStatsV4Payload` and made the `startDate` as optional. So, if no startDate is passed, based on the `StatsGranularity`, the `startDate` would be fetched from `DateUtils`. This was how it is implemented currently. These changes are made in 571764e.
- Also merged with latest changes from `develop` to fix merge conflicts. 

@AmandaRiu could I please trouble you for a review? Thank you!!!

AmandaRiu commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -427,4 +470,121 @@ class WCStatsStore @Inject constructor(
+     * */
+    private fun getRandomPageInt(forced: Boolean): Int {
+        val randomInt = Random.nextInt(STATS_API_V4_MIN_PER_PAGE_PARAM, STATS_API_V4_MAX_PER_PAGE_PARAM)
+        return if (!forced) {
```
np: I think I mentioned this before so this may be a duplicate, but I recommend avoiding negative boolean checks as much as possible for better code readability. They definitely do have their place, but in an `if/else` I try to also handle the boolean as-is if possible instead of doing a `!` check. For example:
```
if (forced) {
    // generate and return new int
} else {
   // ...
}
```

AmandaRiu commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt:
```diff
@@ -251,4 +253,43 @@ object DateUtils {
+     */
+    fun getEndDateForSite(site: SiteModel) = getDateTimeForSite(site, DATE_TIME_FORMAT_END, null)
+
+    fun getStartOfCurrentDay(): String {
```
Shouldn't this be localized for the site timezone?

anitaa1990 replied:
So, we get the start date based on the `StatsGranularity` value and then pass this date string to get localised based on the site timezone [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/7b954f5ce7a5d571227576add64221d8fcc57b69/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt#L514). This is currently how we localise the current date. Instead of the current date, we now pass the `startDate` based on the granularity.

AmandaRiu commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt:
```diff
@@ -251,4 +253,43 @@ object DateUtils {
+        return formatDate(DATE_FORMAT_DEFAULT, cal.time)
+    }
+
+    fun getFirstDayOfCurrentWeek(): String {
```
Shouldn't this be localized for the site timezone?

anitaa1990 replied:
Same comment as above 👍

AmandaRiu commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt:
```diff
@@ -251,4 +253,43 @@ object DateUtils {
+        return formatDate(DATE_FORMAT_DEFAULT, cal.time)
+    }
+
+    fun getFirstDayOfCurrentMonth(): String {
```
Shouldn't this be localized for the site timezone?

anitaa1990 replied:
Same comment as above 👍

AmandaRiu commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/DateUtils.kt:
```diff
@@ -251,4 +253,43 @@ object DateUtils {
+        return formatDate(DATE_FORMAT_DEFAULT, cal.time)
+    }
+
+    fun getFirstDayOfCurrentYear(): String {
```
Shouldn't this be localized for the site timezone?

anitaa1990 replied:
Same comment as above 👍

AmandaRiu commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/orderstats/RevenueStatsApiResponse.kt:
```diff
@@ -0,0 +1,8 @@
+import org.wordpress.android.fluxc.network.Response
+
+class RevenueStatsApiResponse : Response {
+    val intervals: JsonElement? = null
```
The network response provides rolled up stats as well in the `totals` element. I'm curious why we don't include them? May be a lot easier than looping over the data to get the totals later - will definitely be less expensive. We could store the `totals` element as a column in the database.

```
totals: {orders_count: 1, num_items_sold: 1, gross_revenue: 16, coupons: 0, coupons_count: 0, refunds: 0,…}
```

anitaa1990 replied:
I had first added this as a separate column in the db. But then noticed that we would need to fetch the total for revenue and order separately and that would be a total of 4 db separate db calls, even if there are maybe less than 2 orders for a given time period. We _could_ fetch the `WCOrderStatsV4Model` directly from the client and make the required modifications to the data on the client side? WDYT? I think that's what we did for customStats implementation during my trial

anitaa1990 replied:
I've made a sample implementation in this [patch file](https://gist.github.com/anitaa1990/0faf9e0ee27ae3b2f29db32d75bac4ac)

anitaa1990 replied:
Modified the changes in 259c40c

AmandaRiu commented on example/src/main/java/org/wordpress/android/fluxc/example/ui/stats/WooV4StatsFragment.kt:
```diff
@@ -0,0 +1,169 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        stats_select_site.setOnClickListener {
```
I really love that you added a site selector!! <3

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/orderstats/OrderStatsRestClient.kt:
```diff
@@ -45,12 +51,29 @@ class OrderStatsRestClient(
+             *  `This month` tab is selected: [OrderStatsApiUnit] field passed to the API should be [DAY]
+             *  `This year` tab is selected: [OrderStatsApiUnit] field passed to the API should be [MONTH]
+             */
+            fun fromStatsGranularityV4(granularity: StatsGranularity): OrderStatsApiUnit {
```
Thanks @AmandaRiu! I agree. I modified it to `convertToRevenueStatsInterval` though since we are fetching the `interval` field data here from the `StatsGranularity`. Done in ac5fcbd

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/WCStatsV4SqlUtils.kt:
```diff
@@ -0,0 +1,48 @@
+        }
+    }
+
+    fun getRawStatsForSiteIntervalAndDate(
```
👌 Done in 7b954f5

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/persistence/WCStatsV4SqlUtils.kt:
```diff
@@ -0,0 +1,48 @@
+import org.wordpress.android.fluxc.model.WCOrderStatsV4Model
+import org.wordpress.android.fluxc.store.WCStatsStore.StatsGranularity
+
+object WCStatsV4SqlUtils {
```
Done in 7b954f5 👍

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -427,4 +470,121 @@ class WCStatsStore @Inject constructor(
+        }?.toMap() ?: mapOf()
+    }
+
+    private fun getRawStats(
```
👍 Done in 31ee414

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -427,4 +470,121 @@ class WCStatsStore @Inject constructor(
+
+    private fun handleFetchRevenueStatsCompleted(payload: FetchOrderStatsV4ResponsePayload) {
+        val onStatsChanged = with(payload) {
+            if (isError || stats == null) {
```
Ahh, to be honest, I added that because since the `stats` field is an optional value in `FetchOrderStatsV4ResponsePayload`, we would need to do something like `stats?.let{ }` or `stats!!` and the null check seemed like a better option 😄

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -427,4 +470,121 @@ class WCStatsStore @Inject constructor(
+        emitChange(onStatsChanged)
+    }
+
+    fun getRevenueStatsV4(
```
Good point 👍 Done in cee1295

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -427,4 +470,121 @@ class WCStatsStore @Inject constructor(
+        }?.toMap() ?: mapOf()
+    }
+
+    fun getOrderStatsV4(
```
Done in cee1295 👍

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -427,4 +470,121 @@ class WCStatsStore @Inject constructor(
+     * Given a [startDate], formats the date based on the site's timezone in format yyyy-MM-dd'T'hh:mm:ss
+     * If the start date is empty, fetches the date based on the [granularity]
+     */
+    private fun getFormattedStartDate(site: SiteModel, granularity: StatsGranularity, startDate: String?): String {
```
Done in cee1295

anitaa1990 commented on plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -427,4 +470,121 @@ class WCStatsStore @Inject constructor(
+     * So this logic generates a random value between 31 to 100 only if the [forced] is set to true
+     * And storing this value locally to be used when the [forced] flag is set to false.
+     * */
+    private fun getRandomPageInt(forced: Boolean): Int {
```
Done in cee1295 and modified the comments in 65989a4

anitaa1990 commented on plugins/woocommerce/src/main/java/org/wordpress/android/fluxc/action/WCStatsAction.java:
```diff
@@ -16,6 +18,9 @@
     FETCH_ORDER_STATS,
 
+    @Action(payloadType = FetchOrderStatsV4Payload.class)
+    FETCH_ORDER_STATS_V4,
```
That's a good point. I've made modifications to the `WcStatsAction` and renamed the payload classes as well back to `Revenue` in 02746d5. I also reverted the table name and the model name back to `WcRevenueStatsModel` in 7042d35

anitaa1990 commented:
Thanks for the thorough review @AmandaRiu !! I've fixed all the nitpicks, modified the v4 naming changes back to `Revenue` and updated the PR description. PR ready for another round 👍

AmandaRiu commented on example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_WCStatsTest.kt:
```diff
@@ -315,6 +316,159 @@ class MockedStack_WCStatsTest : MockedStack_Base() {
     }
 
+    @Test
+    fun testV4StatsDayFetchSuccess() {
```
np: rename to us 'revenue' for consistency

anitaa1990 replied:
Done in 1233599

AmandaRiu commented on example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_WCStatsTest.kt:
```diff
@@ -315,6 +316,159 @@ class MockedStack_WCStatsTest : MockedStack_Base() {
+    }
+
+    @Test
+    fun testV4StatsFetchCaching() {
```
np: rename to us 'revenue' for consistency

anitaa1990 replied:
Done in 1233599

AmandaRiu commented on example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_WCStatsTest.kt:
```diff
@@ -315,6 +316,159 @@ class MockedStack_WCStatsTest : MockedStack_Base() {
+    }
+
+    @Test
+    fun testV4StatsFetchInvalidParamError() {
```
np: rename to us 'revenue' for consistency

anitaa1990 replied:
Done in 1233599 👍

AmandaRiu commented on example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_WCStatsTest.kt:
```diff
@@ -315,6 +316,159 @@ class MockedStack_WCStatsTest : MockedStack_Base() {
+    }
+
+    @Test
+    fun testV4StatsFetchResponseNullError() {
```
np: rename to us 'revenue' for consistency

anitaa1990 replied:
Done in 1233599

AmandaRiu commented on example/src/main/java/org/wordpress/android/fluxc/example/di/FragmentsModule.kt:
```diff
@@ -60,6 +61,9 @@ internal abstract class FragmentsModule {
     abstract fun provideWooStatsFragmentInjector(): WooStatsFragment
 
+    @ContributesAndroidInjector
+    abstract fun provideWooV4StatsFragmentInjector(): WooV4StatsFragment
```
np: rename to us 'revenue' for consistency

anitaa1990 replied:
Done in b7b01d7

AmandaRiu commented on example/src/main/java/org/wordpress/android/fluxc/example/ui/stats/WooV4StatsFragment.kt:
```diff
@@ -0,0 +1,169 @@
+import org.wordpress.android.fluxc.store.WooCommerceStore
+import javax.inject.Inject
+
+class WooV4StatsFragment : Fragment() {
```
np: rename to us 'revenue' for consistency

anitaa1990 replied:
Done in b7b01d7

AmandaRiu commented on example/src/main/res/layout/fragment_woo_v4_stats.xml:
```diff
@@ -0,0 +1,97 @@
+<ScrollView
```
np: rename file to us 'revenue' for consistency

anitaa1990 replied:
Done in b7b01d7

AmandaRiu commented on example/src/test/java/org/wordpress/android/fluxc/endpoints/WCWPAPIEndpointTest.kt:
```diff
@@ -19,4 +19,9 @@ class WCWPAPIEndpointTest {
     }
+
+    @Test
+    fun testV4StatsUrl() {
```
np: rename to us 'revenue' for consistency

anitaa1990 replied:
Done in d3f46f0

anitaa1990 commented:
@AmandaRiu I've modified all the comments, methods, classes and files from v4 to revenue. I've also added the `total` column to the `WcRevenueStatsModel` so we don't have to calculate the sum during runtime in the app, as you suggested.

PR ready for another round 👍

AmandaRiu commented on example/src/test/java/org/wordpress/android/fluxc/wc/stats/WCStatsSqlUtilsTest.kt:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    @Test
+    fun testSimpleInsertionAndRetrieval() {
```
np: Since this test class will likely be used in the future for other stats endpoints, I recommend renaming these methods so it's clear that they test `revenue` stats

anitaa1990 replied:
Committed in 379651a

AmandaRiu commented on example/src/test/java/org/wordpress/android/fluxc/wc/stats/WCStatsStoreTest.kt:
```diff
@@ -1229,4 +1231,174 @@ class WCStatsStoreTest {
     }
+
+    @Test
+    fun testFetchCurrentDayStatsDate() {
```
np: Since stats can come from different endpoints, I recommend renaming methods relating to `revenue` stats to include `revenue` in the method name.

anitaa1990 replied:
Committed in 379651a 👍

AmandaRiu commented on fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.java:
```diff
@@ -821,18 +832,43 @@ private void migrateAddOn(@AddOn String addOnName, SQLiteDatabase db, int oldDbV
                                + "FOREIGN KEY(LOCAL_SITE_ID) REFERENCES SiteModel(_id) ON DELETE CASCADE,"
                                + "UNIQUE (REMOTE_ORDER_ID, LOCAL_SITE_ID) ON CONFLICT REPLACE)");
                     break;
+                case 75:
```
Since this code has never been released, please remove the previous tables that never made it to production and leave only the table relevant to this PR. It will require us to uninstall and reinstall the app for testing, but that's not a big deal.

anitaa1990 replied:
Thanks for confirming that! Done in 714fb7c

anitaa1990 commented:
@AmandaRiu thanks for the quick review! Updated `WellSqlConfig` to remove all redundant tables created in this PR and fixed the nitpicks. Ready for another round 👍
